crates/driver/src/domain/time.rs

```diff
@@ -0,0 +1,60 @@
+use {
+    crate::infra::{
+        solver::Timeouts,
+        {self},
+    },
+    thiserror::Error,
+};
+
+/// Deadlines for different parts of the driver execution.
+/// The driver is expected to return the solution to the autopilot before the
+/// driver deadline.
+/// The solvers are expected to return the solution to the driver before the
+/// solvers deadline.
+#[derive(Copy, Clone, Debug, Default)]
+pub struct Deadline {
+    driver: chrono::DateTime<chrono::Utc>,
+    solvers: chrono::DateTime<chrono::Utc>,
+}
+
+impl Deadline {
+    pub fn new(deadline: chrono::DateTime<chrono::Utc>, timeouts: Timeouts) -> Self {
+        let deadline = deadline - timeouts.http_delay;
+        Self {
+            driver: deadline,
+            solvers: {
+                let now = infra::time::now();
+                let duration = deadline - now;
```

Comment:
Should we check for negative values here to in order to be robust?

Reply:
No need, negative duration is properly handled in `chrono`, while also we should use the public api of the `Deadline` where we are protected against expired deadlines.